cpp/src/arrow/ipc/metadata.cc

```diff
@@ -591,7 +593,20 @@ static Status SchemaToFlatbuffer(FBB& fbb, const Schema& schema,
     field_offsets.push_back(offset);
   }
 
-  *out = flatbuf::CreateSchema(fbb, endianness(), fbb.CreateVector(field_offsets));
+  /// Custom metadata
+  auto custom_metadata_ = schema.custom_metadata();
```

Comment:
per below, this calls the copy constructor. One way to prevent this is to make KeyValueMetadata non-copyable. If you did that, you would need to pass around something like `std::unique_ptr<const KeyValueMetadata>`. Not sure this is a big deal right now